keps/sig-auth/4193-tokenrequest-requester-info-enhancements/README.md

```diff
@@ -0,0 +1,616 @@
+binding directly to Node objects (without needing to bind to a Pod as well).
+
+This allows users to obtain a token that is tied specifically to the *Node* objects lifecycle, i.e. when the Node
+object is deleted, the token will be invalidated.
+
+### Extending TokenReview to allow cross-checking the embedded Node information with existing Node objects
+
+The SA authenticator will also be extended to check whether a JWT's embedded node information is still valid/current,
+and if not, will reject the review/indicate to the client that the token mismatches with the current state of Nodes.
+
+For tokens bound to Pod objects, to avoid unexpected breaking changes in behaviour, this will not only be protected by
+the feature flag whilst the feature is graduating to GA, but will also be gated behind an additional kube-apiserver flag
+`--service-account-validate-node-info` which defaults to 'off'. This avoids changing the previous behaviour for
+validation of tokens issued for pods.
+
+Tokens that are directly bound to Node objects will always validate the name and UID, as binding tokens to Node objects
+is a new option and therefore enforcing this validation check from day 1 is non-breaking.
+
+The validation will involve first checking whether the Node object with the name given in the JWT still exists, and if
+it does, validating whether the UID of that Node is equal to the UID embedded in the token.
+
+For pods, this behaviour matches that of the existing `pod-garbage-collector-controller`, which will automatically
+delete Pod  objects that are bound to Node's that no longer exist after a period of time. With this enabled, tokens will
+be immediately invalidated rather than waiting until [the 40s grace period](https://github.com/kubernetes/kubernetes/blob/fc786dcd1d2efcc241e0e2392086934f2806555d/pkg/controller/podgc/gc_controller.go#L50-L52) passes.
+
+This means administrators can **delete node objects to invalidate all JWTs issued that embed that Node's info**.
+
+### Including a UUID ([JTI](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.7)) on each issued JWT
+
+When a TokenRequest is being issued/fulfilled, we will modify the issuing code to also generate and embed a UUID which
+can be later used to trace the requests that a specific issued token has made to the apiserver via the audit log.
+
+This will require changing the JWT issuing code to actually generate this UUID, as well as extending the code around the
+audit log to have it record this information into audit entries.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Alice is building an identity issuance system that relies on verifying projected service account tokens to ensure that
+a pod exists, and is associated with a particular service account.
+
+Alice wants to prevent a malicious actor being able to replay this token from another host, so that she can be certain
```

Comment:
This assumes the actual origination IP is exposed to the consumer (not proxy) and no promiscuous links exist, right?  Otherwise IP spoofing would function as expected and via a proxy all origins looks the same.

Reply:
No - the other actor would also need access to the host's certificate in this particular use-case/setup. Origin IP is not used to validate the request. The use of the host's own identity _plus_ a pod that with a claim that references that same node is how we cross-reference and avoid the replay attack.

Perhaps my wording is wrong here, as you are right in saying that if someone had *both* of these documents, they *would* be able to replay the credential.